File: heap_sort/python/heap_sort.py
def heapify(arr, n, var):
    """
    Recursive function to rearrange a heap to maintain the heap property.
    Heap Property:  The key of the root node is greater or less than
                    or equal to the keys of its children.
    :param arr: A list of element to heapify.
    :param n:   Size of heap.
    :param var: To heapify subtree rooted at index.
    """
    highest = var
    left = (2*var)+1
    right = (2*var)+2

    if left < n and arr[var] < arr[left]:
        highest = left

    if right < n and arr[highest] < arr[right]:
        highest = right

    if highest != var:
        arr[var], arr[highest] = arr[highest], arr[var]

        heapify(arr, n, highest)


def heapSort(arr):
    """
    The main function to implement Heap Sorting Algorithm.
    <https://en.wikipedia.org/wiki/Heapsort>

    :param arr: A list of element to sort.
    """
    n = len(arr)
    for i in range(n, -1, -1):
        heapify(arr, n, i)

    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)

    return arr


def main():
    arr = [15, 12, 36, 63, 96]
    print('Sorted elements using Heap Sort: {}'.format(
              ' '.join(map(str, heapSort(arr)))))


if __name__ == '__main__':
    main()

File: heap_sort/cpp/heap_Sort.cpp

#include <iostream>
using namespace std;
void heapify(int arr[], int m, int root)
{
   int largest = root; // root is the largest element
   int l = 2*root + 1; // left = 2*root + 1
   int r = 2*root + 2; // right = 2*root + 2
   // If left child is larger than root
   if (l < m && arr[l] > arr[largest])
   largest = l;// If right child is larger than largest so far
   if (r < m && arr[r] > arr[largest])
   largest = r;// If largest is not root
   if (largest != root)
      {
      swap(arr[root], arr[largest]);//swap root and largest
     heapify(arr, m, largest); // Recursively heapify the sub-tree
      }
}
void heapSort(int arr[], int m)
{

   for (int i = m / 2 - 1; i >= 0; i--)// build heap
   heapify(arr, m, i);
   for (int i=m-1; i>=0; i--)// extracting elements from heap one by one
   {

      swap(arr[0], arr[i]);// Move current root to end
      heapify(arr, i, 0);// again call max heapify on the reduced heap
   }
}

int main()
{
   int m ;
   cout<<"Enter totoal number of elements"<<endl;
   cin>>m;
   int ar[m];
   cout<<"Enter the array"<<endl;
   for(int i=0;i<m;i++)
	{
	   cin>>ar[i];
	}
   heapSort(ar, m);
   cout << "Sorted array is"<<endl;
   for (int i=0; i<m; ++i){
   cout << ar[i] << " ";
   cout << "\n";
   }
}
